File: src/main/java/com/mercateo/jsonschema/property/PropertyDescriptorDefault.java
package com.mercateo.jsonschema.property;

import com.mercateo.immutables.Tuple;
import com.mercateo.jsonschema.generictype.GenericType;
import javaslang.collection.List;
import javaslang.collection.Map;
import javaslang.collection.Set;
import org.immutables.value.Value;

import java.lang.annotation.Annotation;

@Value.Immutable
@Tuple
public abstract class PropertyDescriptorDefault implements PropertyDescriptor {
    public abstract PropertyType propertyType();

    public abstract GenericType<?> genericType();

    public abstract List<Property> children();

    public abstract Map<Class<? extends Annotation>, Set<Annotation>> annotations();
}

File: src/main/java/com/mercateo/jsonschema/property/PropertyDescriptorDefault.kt
package com.mercateo.jsonschema.property

import com.mercateo.jsonschema.generictype.GenericType

data class PropertyDescriptorDefault<T>(
        override val propertyType: PropertyType,
        override val genericType: GenericType<T>,
        override val context: PropertyDescriptor.Context,
        override val annotations: Map<Class<out Annotation>, Set<Annotation>>
) : PropertyDescriptor<T> {
    override val children: List<Property<T, Any>> =
            when (context) {
                is PropertyDescriptor.Context.Children<*> -> context.children as List<Property<T, Any>>
                else -> emptyList()
            }

    val innerReference: Boolean =
            when (context) {
                is PropertyDescriptor.Context.InnerReference -> true
                else -> false
            }
}

